lib/internal/repl.js

```diff
@@ -192,31 +188,49 @@ function setupHistory(repl, historyPath, oldHistoryPath, ready) {
     timer = setTimeout(flushHistory, kDebounceHistoryMS);
   }
 
+  const historyLockFile = `${historyPath}.LCK`;
   function flushHistory() {
     timer = null;
     if (writing) {
       pending = true;
       return;
     }
-    writing = true;
-    const historyData = repl.history.join(os.EOL);
-    fs.write(repl._historyHandle, historyData, 0, 'utf8', onwritten);
-  }
-
-  function onwritten(err, data) {
-    writing = false;
-    if (pending) {
-      pending = false;
-      online();
-    } else {
-      repl._flushing = Boolean(timer);
-      if (!repl._flushing) {
-        repl.emit('flushHistory');
+    // Ref: https://github.com/nodejs/node/issues/1634
+    // opening a file for exclusive write should be atomic
+    // on non-networked POSIX systems.
+    fs.open(historyLockFile, 'wx', (err, fd) => {
+      if (err && err.code === 'EEXIST') {
+        pending = true;
+        repl.resume();
+        return;
       }
-    }
+      writing = true;
+      const lastLine = repl.history[0] + '\n';
+      fs.write(repl._historyHandle, lastLine, NaN, 'utf8', onwritten(fd));
+    });
   }
-}
 
+  function onwritten(fd) {
+    return (err, data) => {
+      writing = false;
+      if (pending) {
+        pending = false;
+        online();
+      } else {
+        repl._flushing = Boolean(timer);
+        if (!repl._flushing) {
+          repl.emit('flushHistory');
+        }
+      }
+      unlockHistoryFile(fd);
```

Comment:
Hmm - I can only reproduce this when there is already a stray `.node_repl_history.LCK` file at start up, and the history file is never written to. When the `REPLServer` is closing it adds a listener for `flushHistory` [here](https://github.com/nodejs/node/blob/master/lib/repl.js#L527-L539) which never gets called, and so the final `Interface#close` is never called.

I can add a check for the existence of the `.LCK` file at startup and then remove it if necessary. There are maybe a couple of ways this could happen.

  * Notify the user. "Hey! There's a lock file, do you want me to delete it?". Probably the safest, but pretty intrusive and kind of looks like its underwear is showing.
  * Set one or more brief timeouts to check/recheck for the existence before deleting. This doesn't eliminate the possibility of a race condition, but it definitely minimizes it.